### PR TITLE
refactor: use Stream instead of Sync methods

### LINKED ISSUE #192: Workflow diff detection KO
Hi @scolladon  :)

Repro branch : issue/192/workflow-rule

Command : sfdx sgd:source:delta -d -t HEAD -f HEAD~1 -r . -i .forceignore -o output

Repro steps : 
Add new fieldUpdates in a WFR and see that output doesnt contains this new diff.

Expected result : 
BE_statut echeance (and other diff) is in output file :
image

Current result : 

BE_statut_echeance (and other diff) are not there

Versions:
Node version : 
+ node --version
v16.10.0
+ echo 'Npm version : '
+ npm --version
Npm version : 
7.24.0
sfdx-git-delta 4.9.0

### PR DESCRIPTION
**Pull request by scolladon**
## What does this pull request contains

---

- [ ] Added for new features.
- [ ] Changed for changes in existing functionality.
- [ ] Deprecated for soon-to-be removed features.
- [ ] Removed for now removed features.
- [X] Fixed for any bug fixes.
- [ ] Security in case of vulnerabilities.

## Explain your changes

---

This refactor allow to treat everything line by line and benefits from the async mechanism of javascript while keeping it easy to read, test and maintain

closes #246 

## Any particular element to being able to test locally

---

Use a very large repository (>20k metadata element)

## Where has this been tested?

---

**Operating System:** Darwin Kernel Version 19.6.0: Sun Nov 14 19:58:51 PST 2021; root:xnu-6153.141.50~1/RELEASE_X86_64

**yarn version:** 1.22.15

**node version:** v14.18.1

**git version:** 2.24.1

**sfdx version:** sfdx-cli/7.132.0 darwin-x64 node-v14.18.1

**sgd plugin version:** 4.12.0

### CODE REVIEW

**Review by arafesthain — COMMENTED** [review]
sounds like "déjà vu".

Je me rappel d'une PR qui enlevait toutes les promises il y a 1 ou 2 ans suite à une issue que j'avais open :D

**Comment by scolladon** [comment]
Yep ! We need to move back to event and async processing to improve buffer and performance...

Thanks to await/async

File: src/utils/cliHelper.js [hunk 1]
```diff
@@ -1,45 +1,81 @@
 'use strict'
 const RepoSetup = require('./repoSetup')
 const { sanitizePath } = require('./childProcessUtils')
-
 const fs = require('fs')
-const git = require('git-state')
+const path = require('path')
+
+const dirExist = async dir => {
```
Comment (arafesthain) [inline src/utils/cliHelper.js:7]: isDir?

File: src/utils/cliHelper.js [hunk 2]
```diff
@@ -1,45 +1,81 @@
...
+  try {
+    const stat = await fs.promises.stat(dir)
+    return stat.isDirectory()
+  } catch {
+    return false
+  }
+}
+const fileExist = async file => {
```
Comment (arafesthain) [inline src/utils/cliHelper.js:15]: isFile?

**Review by mehdisfdc — DISMISSED** [review]
I'm not sure I was able to fully assess the impact of the of the changes, but I did not detect any regression in the manual tests I performed. Looks good to me 🚀

**Review by mehdisfdc — APPROVED** [review]

**Pull request merged**